sarif.fsx

```diff
@@ -0,0 +1,295 @@
+        Decode.object (fun get -> { driver = get.Required.Field "driver" driverDecoder })
+
+    let invocationDecoder: Decoder<Invocation> =
+        Decode.object (fun get ->
+            { startTimeUtc = get.Required.Field "startTimeUtc" Decode.datetimeUtc
+              endTimeUtc = get.Required.Field "endTimeUtc" Decode.datetimeUtc
+              executionSuccessful = get.Required.Field "executionSuccessful" Decode.bool })
+
+    let runDecoder: Decoder<Run> =
+        Decode.object (fun get ->
+            { results = get.Required.Field "results" (Decode.list resultDecoder)
+              tool = get.Required.Field "tool" toolDecoder
+              invocations = get.Required.Field "invocations" (Decode.list invocationDecoder)
+              columnKind = get.Required.Field "columnKind" Decode.string })
+
+    let sarifLogDecoder: Decoder<SarifLog> =
+        Decode.object (fun get ->
+            { schema = get.Required.Field "$schema" Decode.string
+              version = get.Required.Field "version" Decode.string
+              runs = get.Required.Field "runs" (Decode.list runDecoder) })
+
+let private readSarif (json: string) : Result<SarifLog, string> =
+    match Thoth.Json.Newtonsoft.Decode.fromString Decoders.sarifLogDecoder json with
+    | Ok sarifLog -> Ok sarifLog
+    | Error err -> Error($"Failed to decode, got %A{err}")
+
+let private writeSarif (sarifLog: SarifLog) : string =
+    Encoders.sarifLogEncoder sarifLog |> Thoth.Json.Newtonsoft.Encode.toString 4
+
+let mergeSarifFiles _ =
+    task {
+        let mergedPath =
+            Path.Combine(__SOURCE_DIRECTORY__, "analysisreports", "merged.sarif")
+
+        if Path.Exists(mergedPath) then
+            File.Delete(mergedPath)
+
+        let! sarifFiles =
+            Directory.GetFiles("analysisreports", "*.sarif")
+            |> Seq.map (fun path ->
+                task {
+                    let! sarifContent = File.ReadAllTextAsync(path)
+                    let sarifResult = readSarif sarifContent
+
+                    match sarifResult with
+                    | Error e ->
+                        eprintfn $"%A{e}"
+                        return exit 1
+                    | Ok sarif -> return path, sarif
+                })
+            |> Task.WhenAll
+
+        if Array.isEmpty sarifFiles then
+            printfn "No sarif files could be merged"
+        else
+            let firstSarif = snd sarifFiles.[0]
+            let firstRun = firstSarif.runs.[0]
+
+            let results = ResizeArray()
+            let rules = ResizeArray()
+
+            for _, sarif in sarifFiles do
+                for run in sarif.runs do
+                    results.AddRange(run.results)
+
+                    match run.tool.driver.rules with
+                    | None -> ()
+                    | Some rulesList -> rules.AddRange(rulesList)
+
+
+            let combined: SarifLog =
+                {
+                  // I don't know why firstSarif.schema is null
```

Comment:
Maybe something with the `$` 🤷‍♂️ 
But good enough.

Reply:
Looks fine now.